Relaxation boundary for potential

bendudson opened the pull request:
Similar to what is used in STORM, relax the radial boundary value of phi towards zero-gradient on a given timescale.

bendudson committed:
Relaxation boundary for potential

Similar to what is used in STORM, relax the radial boundary value of
phi towards zero-gradient on a given timescale.

johnomotani commented on hermes-2.cxx:
```diff
@@ -1021,6 +1051,77 @@ int Hermes::rhs(BoutReal t) {
+        // Only update if time has advanced
+        // Uses an exponential decay of the weighting of the value in the boundary
+        // so that the solution is well behaved for arbitrary steps
+        BoutReal weight = exp(-(t - phi_boundary_last_update) / phi_boundary_timescale);
```
I don't see an update of `phi_boundary_last_update = t` in this branch. Shouldn't there be one?

Also, if I'm following, this update is done on the first rhs evaluation with an increased `t`, but before calculating `phi`, so that we use the value from the converged time-advance. Doesn't that mean that instead of `(t - phi_boundary_last_update)` we want the length of the _previous_ timestep?

bendudson replied:
Ah yes, sorry good catch! This is what comes of late-night coding. Actually it happens in the day too, so maybe it's something else :). Thanks for the code reviews, it helps a lot!
I hope the way this would work is:
- First rhs call `phi_boundary_last_update` is set to `t`; the boundary is not updated; phi is solved using either the initial condition or value from restart file
- Subsequent calls the difference `t - phi_boundary_last_update` is the time since the last call (the previous timestep); the boundary is updated; phi is solved using the updated boundary

I think this should work as long as `phi_boundary_last_update` is updated after the `weight` factor is calculated.

bendudson committed:
Fix phi boundary relax, update time

When the boundary is modified, the `phi_boundary_last_update`
time needs to be set.

bendudson committed:
Change how phi boundary is set

The INVERT_SET flag takes the value from the guard cell,
then sets this value on the boundary between cells. This results
in an unstable solution if the guard cells are then put back
in as the new boundary value.

With this fix, the old value of the boundary is taken from the
midpoint, and then put into the guard cell. Hopefully this is
consistent with how the value is used in the Laplacian inversion.

johnomotani commented:
> With this fix, the old value of the boundary is taken from the midpoint, and then put into the guard cell. Hopefully this is consistent with how the value is used in the Laplacian inversion.

:+1: should be consistent.

bendudson commented:
Thanks @johnomotani ! I've just realised there's another problem (which was there before): The quantity being solved for is `phi + Pi` rather than `phi`. This shouldn't matter much for the old runs, but I'm finding currently that the value of phi keeps decreasing every iteration, so the time step never converges. Hopefully taking Pi into account will fix it...

bendudson committed:
Include Pi in phi boundary conditions

- Need to move phi boundary values from cell midpoint to boundary cell
  every time the solver is called.
- Need to include Pi in the solve, because Pi is subtracted from the
  result (Boussinesq approximation)

johnomotani reviewed (approved):
LGTM.

Does the `newSolver` branch need to use `phiboundary3d` too to account for `Pi`?

bendudson committed:
Fix phi in outer boundary cell

I don't think this should be necessary, but it seems to be.
Without setting phi in the outermost boundary cells (+2, -2)
the time advance eventually fails to converge. When this failure
occurs all fields seem to be well behaved except these outer boundary
cells which contain large (1e200) values.

With this change no such failures seem to happen (so far)

bendudson committed:
Fix initial guess given to new XZ solver

Initial guess should be phi + Pi, rather than just phi.

bendudson committed:
Always save phi to restart files if evolving vorticity

This is to allow simulations to be run first without relaxing
boundaries, to set the potential to a function of the electron
temperature, and then relaxing boundaries can be turned on.
Not saving potential would mean that relaxing boundaries would restart
with zero potential.

bendudson commented:
I think this is now ready for wider use: It seems to work, though clearly some tests would be nice.
One thing makes me uneasy is that not setting the outermost boundary cells in phi seemed to lead to convergence failures when using relaxing boundaries. Adding two lines to set index ystart-2 and xend+2 in phi stops this from happening. It doesn't seem to affect the values of phi in the domain though...

bshanahan commented:
I've always had this issue with the first and last cells with LaplaceXZ (also with other models).  I don't see it affecting anything in the actual domain though. 

Looks good to me, but I will let John approve this PR as he's been the most active on it (and familiar based on the STORM implementation).

johnomotani commented:
> One thing makes me uneasy is that not setting the outermost boundary cells in phi seemed to lead to convergence failures when using relaxing boundaries. Adding two lines to set index ystart-2 and xend+2 in phi stops this from happening. It doesn't seem to affect the values of phi in the domain though...

very strange! I think it's possible to set a watch point with gdb on the memory locations of one of those xstart-2 array entries, and see if/where it ever gets read?

bendudson committed:
Don't calculate nu if currents are turned off

Not needed, and now results in a warning about missing boundary
conditions if run with currents turned off.

johnomotani commented on hermes-2.cxx:
```diff
@@ -1038,33 +1161,68 @@ int Hermes::rhs(BoutReal t) {
           if (newXZsolver) {
             newSolver->setCoefs(1. / SQ(coord->Bxy), 0.0);
-            phi = newSolver->solve(Vort - Vort2D, phi);
+            phi = newSolver->solve(Vort - Vort2D, phi + Pi); // Second argument is initial guess
```
Shouldn't the second argument be something like `phi_boundary3d - phi_boundary2d + Pi` for the boundary conditions? Although maybe the `- phi_boundary2d` doesn't matter if the solver takes care of filtering the n=0 mode?

If this is used as an initial guess for an iterative solver shouldn't it be `phi - phi2D + Pi`, with a previous call to `phi.setBoundaryTo(phi_boundary3d)` for the bc?

bendudson replied:
Thanks @johnomotani ! Yes, I've modified these `newSolver` branches to (I think) set the initial guess and boundaries correctly. Not tested though...

johnomotani replied:
looks good to me :+1:

johnomotani commented on hermes-2.cxx:
```diff
@@ -1080,7 +1238,7 @@ int Hermes::rhs(BoutReal t) {
           } else {
```
Should the line above (l.1237, but Github won't let me comment on it because it's not close enough to a changed line) be `phi_boundary3d`, or have a call to `phi.setBoundaryTo(phi_boundary3d)` before it?

bendudson committed:
Set boundary for LaplaceXZ solver

Define a function `withBoundary` which sets the boundary.
Use this to set the boundary when using new LaplaceXZ solver.
